## Analyzing New York's Yellow taxi dataset

- The datasets can be accessed from [here](https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page)
- Time period: `January 2022 - December 2025`
- visit [`here`](https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page) to know what each column means
- Dataset Characteristics <br>
 -> File Format: Parquet  
 -> Dataset type: Tabular  
 -> Number of columns: 19. 
 -> Time Period: 4 years (January - December)
-> 2022:
-> 2023:
-> 2024:
-> 2025: 

In [1]:
import pandas as pd

In [2]:
df = pd.read_parquet('../raw/2022/yellow_tripdata_2022-01.parquet')

In [3]:
df

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee
0,1,2022-01-01 00:35:40,2022-01-01 00:53:29,2.0,3.80,1.0,N,142,236,1,14.50,3.0,0.5,3.65,0.0,0.3,21.95,2.5,0.0
1,1,2022-01-01 00:33:43,2022-01-01 00:42:07,1.0,2.10,1.0,N,236,42,1,8.00,0.5,0.5,4.00,0.0,0.3,13.30,0.0,0.0
2,2,2022-01-01 00:53:21,2022-01-01 01:02:19,1.0,0.97,1.0,N,166,166,1,7.50,0.5,0.5,1.76,0.0,0.3,10.56,0.0,0.0
3,2,2022-01-01 00:25:21,2022-01-01 00:35:23,1.0,1.09,1.0,N,114,68,2,8.00,0.5,0.5,0.00,0.0,0.3,11.80,2.5,0.0
4,2,2022-01-01 00:36:48,2022-01-01 01:14:20,1.0,4.30,1.0,N,68,163,1,23.50,0.5,0.5,3.00,0.0,0.3,30.30,2.5,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2463926,2,2022-01-31 23:36:53,2022-01-31 23:42:51,NaN,1.32,NaN,NaN,90,170,0,8.00,0.0,0.5,2.39,0.0,0.3,13.69,NaN,NaN
2463927,2,2022-01-31 23:44:22,2022-01-31 23:55:01,NaN,4.19,NaN,NaN,107,75,0,16.80,0.0,0.5,4.35,0.0,0.3,24.45,NaN,NaN
2463928,2,2022-01-31 23:39:00,2022-01-31 23:50:00,NaN,2.10,NaN,NaN,113,246,0,11.22,0.0,0.5,2.00,0.0,0.3,16.52,NaN,NaN
2463929,2,2022-01-31 23:36:42,2022-01-31 23:48:45,NaN,2.92,NaN,NaN,148,164,0,12.40,0.0,0.5,0.00,0.0,0.3,15.70,NaN,NaN


In [4]:
df.describe()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee
count,2.463931e+06,2463931,2463931,2.392428e+06,2.463931e+06,2.392428e+06,2.463931e+06,2.463931e+06,2.463931e+06,2.463931e+06,2.463931e+06,2.463931e+06,2.463931e+06,2.463931e+06,2.463931e+06,2.463931e+06,2.392428e+06,2.392428e+06
mean,1.707819e+00,2022-01-17 01:19:51.689724,2022-01-17 01:34:04.421902,1.389453e+00,5.372751e+00,1.415507e+00,1.660768e+02,1.635807e+02,1.194449e+00,1.294648e+01,1.006710e+00,4.914539e-01,2.385696e+00,3.749773e-01,2.967234e-01,1.916937e+01,2.282322e+00,8.249935e-02
min,1.000000e+00,2008-12-31 22:23:09,2008-12-31 23:06:56,0.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,-4.800000e+02,-4.500000e+00,-5.000000e-01,-1.252200e+02,-3.140000e+01,-3.000000e-01,-4.803000e+02,-2.500000e+00,-1.250000e+00
25%,1.000000e+00,2022-01-09 15:37:41,2022-01-09 15:50:50.500000,1.000000e+00,1.040000e+00,1.000000e+00,1.320000e+02,1.130000e+02,1.000000e+00,6.500000e+00,0.000000e+00,5.000000e-01,7.200000e-01,0.000000e+00,3.000000e-01,1.130000e+01,2.500000e+00,0.000000e+00
50%,2.000000e+00,2022-01-17 12:11:45,2022-01-17 12:23:49,1.000000e+00,1.740000e+00,1.000000e+00,1.620000e+02,1.620000e+02,1.000000e+00,9.000000e+00,5.000000e-01,5.000000e-01,2.000000e+00,0.000000e+00,3.000000e-01,1.469000e+01,2.500000e+00,0.000000e+00
75%,2.000000e+00,2022-01-24 13:49:37.500000,2022-01-24 14:02:51,1.000000e+00,3.130000e+00,1.000000e+00,2.340000e+02,2.360000e+02,1.000000e+00,1.400000e+01,2.500000e+00,5.000000e-01,3.000000e+00,0.000000e+00,3.000000e-01,2.002000e+01,2.500000e+00,0.000000e+00
max,6.000000e+00,2022-05-18 20:41:57,2022-05-18 20:47:45,9.000000e+00,3.061593e+05,9.900000e+01,2.650000e+02,2.650000e+02,5.000000e+00,4.010923e+05,3.350000e+01,1.659000e+01,8.888800e+02,1.933000e+02,3.000000e-01,4.010956e+05,2.500000e+00,1.250000e+00
std,5.021375e-01,NaN,NaN,9.829686e-01,5.478714e+02,5.917573e+00,6.546806e+01,7.079016e+01,5.001778e-01,2.558149e+02,1.236226e+00,8.291560e-02,2.830698e+00,1.680538e+00,4.374741e-02,2.559641e+02,7.432040e-01,3.125554e-01


In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2463931 entries, 0 to 2463930
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     str           
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            float64   

## Columns and Defenitions

* `VendorID`: A code indicating the TPEP provider that provided the record. <br>
-> 1 = Creative Mobile Technologies, LLC  
-> 2 = Curb Mobility, LLC  
-> 6 = Myle Technologies Inc  
-> 7 = Helix  

* `tpep_pickup_datetime`: The date and time when the meter was engaged.
* `tpep_dropoff_datetime`: The date and time when the meter was disengaged.
* `passenger_count`: The number of passengers in the vehicle.
* `trip_distance`: The elapsed trip distance in miles reported by the taximeter.
* `RatecodeID`: The final rate code in effect at the end of the trip. <br>
-> 1 = Standard rate  
-> 2 = JFK  
-> 3 = Newark  
-> 4 = Nassau or Westchester  
-> 5 = Negotiated fare  
-> 6 = Group ride  
-> 99 = Null/unknown  
* `store_and_fwd_flag`: This flag indicates whether the trip record was held in vehicle memory before
sending to the vendor, aka “store and forward,” because the vehicle did not have a connection to the server. <br>
-> Y = store and forward trip  
-> N = not a store and forward trip  
* `PULocationID`: TLC Taxi Zone in which the taximeter was engaged.
* `DOLocationID`: TLC Taxi Zone in which the taximeter was disengaged.
* `payment_type`: A numeric code signifying how the passenger paid for the trip. <br>
-> 0 = Flex Fare trip  
-> 1 = Credit card  
-> 2 = Cash  
-> 3 = No charge  
-> 4 = Dispute  
-> 5 = Unknown  
-> 6 = Voided trip  
* `fare_amount`: The time-and-distance fare calculated by the meter.
* `extra`: Miscellaneous extras and surcharges.
* `mta_tax`: Tax that is automatically triggered based on the metered rate in use.
* `tip_amount`: Tip amount – This field is automatically populated for credit card tips. Cash tips are not included.
* `tolls_amount`: Total amount of all tolls paid in trip
* `improvement_surcharge`: Improvement surcharge assessed trips at the flag drop
* `total_amount`: The total amount charged to passengers. Does not include cash tips. (Total amount is the sum of `fare_amount`, `extra`, `mta_tax`, `tip_amount`, `tolls_amount`, and `improvement_surcharge`)
* `congestion_surcharge`: Total amount collected in trip for NYS congestion surcharge.
* `airport_fee`: For pick up only at LaGuardia and John F. Kennedy Airports
* `cbd_congestion_fee`: Per-trip charge for MTA's Congestion Relief Zone starting Jan. 5, 2025.

In [6]:
df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee
0,1,2022-01-01 00:35:40,2022-01-01 00:53:29,2.0,3.80,1.0,N,142,236,1,14.5,3.0,0.5,3.65,0.0,0.3,21.95,2.5,0.0
1,1,2022-01-01 00:33:43,2022-01-01 00:42:07,1.0,2.10,1.0,N,236,42,1,8.0,0.5,0.5,4.00,0.0,0.3,13.30,0.0,0.0
2,2,2022-01-01 00:53:21,2022-01-01 01:02:19,1.0,0.97,1.0,N,166,166,1,7.5,0.5,0.5,1.76,0.0,0.3,10.56,0.0,0.0
3,2,2022-01-01 00:25:21,2022-01-01 00:35:23,1.0,1.09,1.0,N,114,68,2,8.0,0.5,0.5,0.00,0.0,0.3,11.80,2.5,0.0
4,2,2022-01-01 00:36:48,2022-01-01 01:14:20,1.0,4.30,1.0,N,68,163,1,23.5,0.5,0.5,3.00,0.0,0.3,30.30,2.5,0.0


## Using DuckDB for quering through the data

In [7]:
import duckdb

In [8]:
df = duckdb.read_parquet('../raw/*/*.parquet')

In [9]:
df.columns

['VendorID',
 'tpep_pickup_datetime',
 'tpep_dropoff_datetime',
 'passenger_count',
 'trip_distance',
 'RatecodeID',
 'store_and_fwd_flag',
 'PULocationID',
 'DOLocationID',
 'payment_type',
 'fare_amount',
 'extra',
 'mta_tax',
 'tip_amount',
 'tolls_amount',
 'improvement_surcharge',
 'total_amount',
 'congestion_surcharge',
 'airport_fee']

In [10]:
df = duckdb.sql("""
                SELECT * EXCLUDE (
                VendorID,
                store_and_fwd_flag
                )
                FROM df
                """)

In [11]:
df

┌──────────────────────┬───────────────────────┬─────────────────┬───────────────┬────────────┬──────────────┬──────────────┬──────────────┬─────────────┬────────┬─────────┬────────────┬──────────────┬───────────────────────┬──────────────┬──────────────────────┬─────────────┐
│ tpep_pickup_datetime │ tpep_dropoff_datetime │ passenger_count │ trip_distance │ RatecodeID │ PULocationID │ DOLocationID │ payment_type │ fare_amount │ extra  │ mta_tax │ tip_amount │ tolls_amount │ improvement_surcharge │ total_amount │ congestion_surcharge │ airport_fee │
│      timestamp       │       timestamp       │     double      │    double     │   double   │    int64     │    int64     │    int64     │   double    │ double │ double  │   double   │    double    │        double         │    double    │        double        │   double    │
├──────────────────────┼───────────────────────┼─────────────────┼───────────────┼────────────┼──────────────┼──────────────┼──────────────┼─────────────┼────────┼───

## Finding negative values in numeric fields

In [12]:
duckdb.sql("""
            SELECT 
            COUNT (*) AS total_rows,
            COUNT (*) FILTER (WHERE trip_distance < 0) AS negative_distance,
            COUNT (*) FILTER (WHERE fare_amount < 0) AS negative_fare,
            COUNT (*) FILTER (WHERE passenger_count < 0) AS negative_passengers,
            COUNT (*) FILTER (WHERE extra < 0) AS negative_extra,
            COUNT (*) FILTER (WHERE mta_tax < 0) AS negative_mta_tax,
            COUNT (*) FILTER (WHERE tip_amount < 0) AS negative_tip_amount,
            COUNT (*) FILTER (WHERE tolls_amount < 0) AS negative_tolls_amount,
            COUNT (*) FILTER (WHERE improvement_surcharge < 0) AS negative_improvement_surcharge,
            COUNT (*) FILTER (WHERE total_amount < 0) AS negative_total_amount,
            COUNT (*) FILTER (WHERE congestion_surcharge < 0) AS negative_congestion_surcharge,
            COUNT (*) FILTER (WHERE airport_fee < 0) AS negative_airport_fee
            FROM df
""").df()

,total_rows,negative_distance,negative_fare,negative_passengers,negative_extra,negative_mta_tax,negative_tip_amount,negative_tolls_amount,negative_improvement_surcharge,negative_total_amount,negative_congestion_surcharge,negative_airport_fee
0,167858646,0,4214194,0,1029414,1958014,8229,158379,2042089,2215654,1639288,344818


## Negative values

- Negative fare: `4214194`
- Negative extra: `1029414`
- Negative mta tax: `1958014`
- Negative tip amount: `8229`
- Negative tolls amount: `158379`
- Negative improvement surcharge: `2042089`
- Negative total amount: `2215654`
- Negative congestion surcharge: `1639288`
- Negative airport fee: `344818`

In [13]:
duckdb.sql("""
            SELECT * 
            FROM df
            WHERE 
            fare_amount < 0 
            
            """)

┌──────────────────────┬───────────────────────┬─────────────────┬───────────────┬────────────┬──────────────┬──────────────┬──────────────┬─────────────┬────────┬─────────┬────────────┬──────────────┬───────────────────────┬──────────────┬──────────────────────┬─────────────┐
│ tpep_pickup_datetime │ tpep_dropoff_datetime │ passenger_count │ trip_distance │ RatecodeID │ PULocationID │ DOLocationID │ payment_type │ fare_amount │ extra  │ mta_tax │ tip_amount │ tolls_amount │ improvement_surcharge │ total_amount │ congestion_surcharge │ airport_fee │
│      timestamp       │       timestamp       │     double      │    double     │   double   │    int64     │    int64     │    int64     │   double    │ double │ double  │   double   │    double    │        double         │    double    │        double        │   double    │
├──────────────────────┼───────────────────────┼─────────────────┼───────────────┼────────────┼──────────────┼──────────────┼──────────────┼─────────────┼────────┼───

## Finding average passengers

In [19]:
duckdb.sql("""
            SELECT 
            AVG(trip_distance) AS average_trip_distance,
            MIN(trip_distance) AS minimum_trip_distance,
            MAX(trip_distance) AS maximum_trip_distance,
            PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY trip_distance) AS p25,
            PERCENTILE_CONT(0.50) WITHIN GROUP (ORDER BY trip_distance) AS p50,
            PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY trip_distance) AS p75,
            PERCENTILE_CONT(0.99) WITHIN GROUP (ORDER BY trip_distance) AS P99
            FROM df 
            """)

┌───────────────────────┬───────────────────────┬───────────────────────┬────────┬────────┬────────┬────────┐
│ average_trip_distance │ minimum_trip_distance │ maximum_trip_distance │  p25   │  p50   │  p75   │  P99   │
│        double         │        double         │        double         │ double │ double │ double │ double │
├───────────────────────┼───────────────────────┼───────────────────────┼────────┼────────┼────────┼────────┤
│     5.547781812859816 │                   0.0 │             398608.62 │   1.05 │   1.81 │    3.5 │   20.0 │
└───────────────────────┴───────────────────────┴───────────────────────┴────────┴────────┴────────┴────────┘

- The above analysis shows that 99 percent of the trips are under 20 miles.
- the maximum trip distance is 398608.62 (outlier)

In [30]:
duckdb.sql("""
            SELECT
            *
            FROM df
            WHERE trip_distance > 100
            """)

┌──────────────────────┬───────────────────────┬─────────────────┬───────────────┬────────────┬──────────────┬──────────────┬──────────────┬─────────────┬────────┬─────────┬────────────┬──────────────┬───────────────────────┬──────────────┬──────────────────────┬─────────────┐
│ tpep_pickup_datetime │ tpep_dropoff_datetime │ passenger_count │ trip_distance │ RatecodeID │ PULocationID │ DOLocationID │ payment_type │ fare_amount │ extra  │ mta_tax │ tip_amount │ tolls_amount │ improvement_surcharge │ total_amount │ congestion_surcharge │ airport_fee │
│      timestamp       │       timestamp       │     double      │    double     │   double   │    int64     │    int64     │    int64     │   double    │ double │ double  │   double   │    double    │        double         │    double    │        double        │   double    │
├──────────────────────┼───────────────────────┼─────────────────┼───────────────┼────────────┼──────────────┼──────────────┼──────────────┼─────────────┼────────┼───

## Extracting Trip Distance Time

In [33]:
duckdb.sql("""
            SELECT 
                trip_distance,
                EXTRACT(EPOCH FROM (
                    tpep_dropoff_datetime - tpep_pickup_datetime
                    )) / 60 AS trip_duration_minutes,
                    fare_amount,
                    total_amount
            FROM df
            WHERE trip_distance > 20
            ORDER BY trip_distance DESC
            """).df()

,trip_distance,trip_duration_minutes,fare_amount,total_amount
0,398608.62,14.000000,22.04,26.04
1,397994.37,18.000000,21.97,23.47
2,389678.46,13.000000,12.11,18.83
3,386088.43,20.000000,17.16,21.91
4,366343.04,35.000000,22.06,26.06
...,...,...,...,...
1677976,20.01,34.400000,95.80,122.63
1677977,20.01,72.300000,88.40,109.63
1677978,20.01,50.900000,70.00,99.23
1677979,20.01,38.566667,97.20,130.58


## References

- https://www.nyc.gov/assets/tlc/downloads/pdf/data_dictionary_trip_records_yellow.pdf
- https://duckdb.org/docs/current/clients/python/overview